### PR TITLE
Issue #8132: Add NoCodeInFileCheck

### LINKED ISSUE #8132: new Cheeck: NoCodeInFile
to cover functionality that we removed from PackgeDeclaration Check at https://github.com/checkstyle/checkstyle/issues/8110

We need new Check to cover cases where all content is commented out. So file is not empty, but there is no java code .

It should print violations on files like the following:
````
$ cat TestClass.java

$ cat TestClass2.java
// violation

$ cat TestClass3.java
/* violation */
````

### PR DESCRIPTION
**Pull request by gaurabdg**
Resolves #8132 

Diff Report: https://gaurabdg.github.io/checkstyle-tester-reports/regression/NoCodeInFile/diff/
Contribution PR: https://github.com/checkstyle/contribution/pull/468

### CODE REVIEW

File: config/checkstyle_non_main_files_suppressions.xml [hunk 1]
```diff
@@ -164,4 +164,6 @@
               files="src[\\/]test[\\/]resources[\\/].*[\\/]regexponfilename[\\/].*"/>
     <suppress checks="RegexpMultiline"
               files="src[\\/]test[\\/]resources[\\/].*[\\/]grammar[\\/]InputRegressionEmptyAst\.txt"/>
+        <suppress checks="RegexpMultiline"
```
Comment (strkkk) [inline config/checkstyle_non_main_files_suppressions.xml:167]: indentation

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+ * &lt;module name="NoCodeInFile"/&gt;
+ * </pre>
+ * <p>
+ * Example:
+ * </p>
+ * <pre>
+ * Files which are considered to have no code:
+ * 1. File with no text
+ * 2. File with single line comment(s).
+ * 3. File with a multi line comment(s).
```
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java:44]: above lines should be in description, here should be examples

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheckTest.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        };
+        verify(checkConfig, getPath("InputNoCodeInFile3.java"), expected);
+    }
+
+    @Test
+    public void testFileContainingCode() throws Exception {
+        final DefaultConfiguration checkConfig = createModuleConfig(NoCodeInFileCheck.class);
+        verify(checkConfig, getPath("InputNoCodeInFile4.java"),
+                CommonUtil.EMPTY_STRING_ARRAY);
+    }
```
Comment (strkkk) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheckTest.java:84]: can be added one more test with both single and multiline comments

File: src/test/java/com/puppycrawl/tools/checkstyle/DetailAstImplTest.java [hunk 4]
```diff
@@ -510,7 +510,10 @@ else if (file.getName().endsWith(".java")
                     && !file.getName().endsWith("InputGrammar.java")
                     // comment only files, no root
                     && !file.getName().endsWith("InputPackageDeclarationWithCommentOnly.java")
-                    && !file.getName().endsWith("InputSingleSpaceSeparatorEmpty.java")) {
+                    && !file.getName().endsWith("InputSingleSpaceSeparatorEmpty.java")
+                    && !file.getName().endsWith("InputNoCodeInFile1.java")
+                    && !file.getName().endsWith("InputNoCodeInFile2.java")
+                    && !file.getName().endsWith("InputNoCodeInFile3.java")) {
```
Comment (strkkk) [inline src/test/java/com/puppycrawl/tools/checkstyle/DetailAstImplTest.java:516]: it is better to extract these endings into some list and iterate over it, it will be more readable

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+package com.puppycrawl.tools.checkstyle.checks;
+
+import com.puppycrawl.tools.checkstyle.StatelessCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
+
+/**
+ * <p>
+ * Checks whether files contain code.
```
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java:29]: `files contain` may be `file contains` is more precise.

File: config/checkstyle_non_main_files_suppressions.xml (hunk 1 above)
Comment (gaurabdg) [reply config/checkstyle_non_main_files_suppressions.xml:167]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java (hunk 5 above)
Comment (gaurabdg) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java:29]: Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/DetailAstImplTest.java (hunk 4 above)
Comment (gaurabdg) [reply src/test/java/com/puppycrawl/tools/checkstyle/DetailAstImplTest.java:516]: Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheckTest.java (hunk 3 above)
Comment (gaurabdg) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheckTest.java:84]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java (hunk 2 above)
Comment (gaurabdg) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java:44]: Done.

**Comment by strkkk** [comment]
@gaurabdg I see your "done" responses but I see no code changes

**Comment by gaurabdg** [comment]
> @gaurabdg I see your "done" responses but I see no code changes

Sorry for the confusion, I was just commenting for me to keep track, I've submitted the new commit.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+import com.puppycrawl.tools.checkstyle.StatelessCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
+
+/**
+ * <p>
+ * Checks whether file contains code. Files which are considered to have no code: 1. File with no
+ * text 2. File with single line comment(s). 3. File with a multi line comment(s).
```
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java:30]: please make list like `<ul>` and items in `<li>` tag

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+ * &lt;module name="NoCodeInFile"/&gt;
+ * </pre>
+ * <p>
+ * Example:
+ * </p> Violations:
+ * <pre>
+ * </pre>
+ * <pre>
+ * // comment
+ * </pre>
```
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java:45]: Please change examples.
Make it like
```
Content of the file:
... // violation
```

File: src/xdocs/config_misc.xml [hunk 8]
```diff
@@ -1541,6 +1541,72 @@ This is a sample text file. // ok, this file is not specified in the config.
...
+            Checkstyle Style</a>
+          </li>
+        </ul>
+      </subsection>
+
+      <subsection name="Violation Messages" id="NoCodeInFile_Violation_Messages">
+        <ul>
+          <li>
+            <a
+                    href="https://github.com/search?q=path%3Asrc%2Fmain%2Fresources%2Fcom%2Fpuppycrawl%2Ftools%2Fcheckstyle%2Fchecks+filename%3Amessages*.properties+repo%3Acheckstyle%2Fcheckstyle+%22nocode.in.file%22">
```
[the remark this answers is not in the document]
Comment (gaurabdg) [reply src/xdocs/config_misc.xml]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java (hunk 6 above)
Comment (gaurabdg) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java (hunk 7 above)
Comment (gaurabdg) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java:59]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+ * &lt;module name="NoCodeInFile"/&gt;
+ * </pre>
+ * <p>
+ * Example:
+ * </p>
+ * <p>
+ * Content of the files:
+ * </p>
+ * <pre>
+ * // violation
```
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java:55]: please change it to block comment example. Showing empty file looks strange

File: src/xdocs/config_misc.xml [hunk 10]
```diff
@@ -1541,6 +1541,83 @@ This is a sample text file. // ok, this file is not specified in the config.
...
+        </p>
+        <ul>
+            <li>
+                File with no text
+            </li>
+            <li>
+                File with single line comment(s)
+            </li>
+            <li>
+                File with a multi line comment(s).
```
Comment (strkkk) [inline src/xdocs/config_misc.xml:1560]: indentation here and above should be 2 spaces

File: src/xdocs/config_misc.xml [hunk 11]
```diff
@@ -1541,6 +1541,83 @@ This is a sample text file. // ok, this file is not specified in the config.
...
+          To configure the check:
+        </p>
+        <source>
+&lt;module name=&quot;NoCodeInFile&quot;/&gt;
+        </source>
+        <p>
+            Example:
+        </p>
+        <p>
+           Content of the files:
```
Comment (strkkk) [inline src/xdocs/config_misc.xml:1576]: same here and above

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java (hunk 9 above)
Comment (gaurabdg) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java:55]: But, in block comment, `*/` is being treated as javadoc end. Anyway to escape it?

Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java:55]: Yes, https://stackoverflow.com/a/631839/3901853

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+ * <p>
+ * Content of the files:
+ * </p>
+ * <pre>
+ * // single line comment // violation
+ * </pre>
+ * <pre>
+ * /*
+ *  block comment
+ * &#42;&#47; // violation
```
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java:60]: no need to escape asterisk

File: src/xdocs/config_misc.xml (hunk 11 above)
Comment (strkkk) [reply src/xdocs/config_misc.xml:1576]: not corrected

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java (hunk 12 above)
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java:60]: violation is reported on line 1, so violation comment should be there as well

**Comment by gaurabdg** [comment]
> @gaurabdg CI is failing again

@strkkk It's strange. The IDEA inspection errors are from files that I haven't modified.

**Comment by strkkk** [comment]
@gaurabdg please try to rebase on latest master

**Comment by gaurabdg** [comment]
> @gaurabdg please try to rebase on latest master

It is on the latest master.

**Comment by strkkk** [comment]
@gaurabdg  TC issue has been fixed, please rebase

**Comment by pbludov** [comment]
@gaurabdg Is it possible to add a test case with zero-length file? The `javac` is fine with them:
```
touch TestClass.java
javac TestClass.java
```
(no errors)

**Comment by gaurabdg** [comment]
> @gaurabdg Is it possible to add a test case with zero-length file? The `javac` is fine with them:
> 
> ```
> touch TestClass.java
> javac TestClass.java
> ```
> 
> (no errors)

@pbludov done.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheckTest.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+
+    @Override
+    protected String getPackageLocation() {
+        return "com/puppycrawl/tools/checkstyle/checks/nocodeinfile";
+    }
+
+    @Test
+    public void testGetRequiredTokens() {
+        final NoCodeInFileCheck checkObj = new NoCodeInFileCheck();
+        assertArrayEquals(CommonUtil.EMPTY_INT_ARRAY, checkObj.getRequiredTokens(),
```
Comment (pbludov) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheckTest.java:41]: Please use google.Truth here
```
assertWithMessage("Required tokens array is not empty")
    .that(checkObj.getRequiredTokens())
    .isEmpty();
```

File: src/xdocs/config_misc.xml [hunk 14]
```diff
@@ -1541,6 +1541,85 @@ This is a sample text file. // ok, this file is not specified in the config.
       </subsection>
     </section>
 
+    <section name="NoCodeInFile">
+      <p>Since Checkstyle 8.33</p>
+
+      <subsection name="Description" id="NoCodeInFile_Description">
+        <p>
+          Checks whether file contains code.
```
Comment (pbludov) [inline src/xdocs/config_misc.xml:1549]: The description of the check is confusing. Will any file with a single/multi-line comment be considered empty? Or just a file with there is nothing except for some comments within?
Could you please make it more clear?

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheckTest.java (hunk 13 above)
Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheckTest.java:41]: and same for similar cases in this file

File: src/xdocs/config_misc.xml (hunk 14 above)
Comment (gaurabdg) [reply src/xdocs/config_misc.xml:1549]: I have specified the conditions in `Files which are considered to have no code:`.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheckTest.java (hunk 13 above)
Comment (gaurabdg) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheckTest.java:41]: Done

**Comment by rnveach** [comment]
CI needs to pass for me to review. Pitest is failing.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XpathRegressionTest.java [hunk 15]
```diff
@@ -59,6 +59,7 @@
             "Indentation",
             "JavadocMethod",
             "MissingJavadocType",
+            "NoCodeInFile",
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/XpathRegressionTest.java:62]: Look at top text, it says are "not compatible ... till ....". This check will never be compatible. It always logs by line number since there is never any AST.

We need to follow pattern of other checks where we couldn't support them, `Regexp` and `RegexpSinglelineJava` and give a reason. xdoc should also be updated too.

Comment (gaurabdg) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XpathRegressionTest.java]: Done.
